packages/engine/lib/error/src/iter.rs

```diff
@@ -0,0 +1,70 @@
+use core::{fmt, fmt::Formatter, iter::FusedIterator, marker::PhantomData};
+
+use provider::TypeTag;
+
+use super::Frame;
+use crate::{tags::FrameSource, Frames, Report, Requests};
+
+impl<'r> Frames<'r> {
+    pub(super) const fn new<C>(report: &'r Report<C>) -> Self {
+        Self {
+            current: Some(&report.inner.frame),
+        }
+    }
+}
+
+impl<'r> Iterator for Frames<'r> {
+    type Item = &'r Frame;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.current.take().map(|current| {
+            self.current = current.request::<FrameSource>();
+            current
+        })
+    }
+}
+
+impl<'r> FusedIterator for Frames<'r> {}
+
+impl fmt::Debug for Frames<'_> {
+    fn fmt(&self, fmt: &mut Formatter<'_>) -> fmt::Result {
+        fmt.debug_list().entries(self.clone()).finish()
+    }
+}
+
+impl<'r, I> Requests<'r, I> {
+    pub(super) const fn new<S>(report: &'r Report<S>) -> Self {
```

Comment:
Any reason for why `<S>` rather than `<C>`?

```suggestion
    pub(super) const fn new<C>(report: &'r Report<C>) -> Self {
```